phoenix-core/src/main/java/org/apache/phoenix/expression/BlobExpression.java

```diff
@@ -0,0 +1,210 @@
+import org.apache.phoenix.schema.types.LobStoreException;
+import org.apache.phoenix.schema.types.PBlob;
+import org.apache.phoenix.util.EnvironmentEdgeManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInput;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutput;
+import java.io.ObjectOutputStream;
+import java.io.Serializable;
+import java.util.UUID;
+
+import static org.apache.phoenix.expression.BlobExpression.BlobMetaData.serializeBlobMetaData;
+
+/**
+ * Expression accessor for BLOB values
+ */
+public class BlobExpression extends LiteralExpression {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(BlobExpression.class);
+    public enum SUPPORTED_FORMATS {
+        S3, HDFS, LOCAL_FS
+    };
+    private final LobStore lobStore;
+
+    public BlobExpression(Object value, SortOrder sortOrder, Determinism determinism) {
+        super(value, sortOrder, determinism, PBlob.INSTANCE);
+        // TODO: hard-coded to fs-based lob store for now. Can be a client-side config/SPI
+        this.lobStore = new FileSystemBasedLobStore();
+    }
+
+    @Override
+    public boolean evaluate(Tuple tuple, ImmutableBytesWritable ptr) {
+        String lobLocation;
+        try {
+            lobLocation = this.lobStore.putLob((InputStream)this.getValue());
+        } catch (LobStoreException lobEx) {
+            throw new RuntimeException("Persisting the BLOB failed", lobEx);
+        }
+        BlobMetaData metaData = getBlobMetaData(true,
+                lobLocation,
+                EnvironmentEdgeManager.currentTimeMillis(),
+                SUPPORTED_FORMATS.LOCAL_FS,
+                UUID.randomUUID());
+        byte[] serializedBlobMetaData;
+        try {
+            serializedBlobMetaData = serializeBlobMetaData(metaData);
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to serialize BLOB metadata", e);
+        }
+        ptr.set(serializedBlobMetaData);
+        return true;
+    }
+
+    public InputStream getInputStream(BlobMetaData blobMetaData) {
+        Preconditions.checkNotNull(blobMetaData);
+        InputStream res;
+        try {
+            res = this.lobStore.getLob(blobMetaData.getLobLocator());
+        } catch (LobStoreException lobEx) {
+            LOGGER.error("Getting BLOB failed", lobEx);
+            return null;
+        }
+        return res;
+    }
+
+    @VisibleForTesting
+    BlobMetaData getBlobMetaData(boolean onDisk, String lobLocation, long lastUpdatedTime,
+            SUPPORTED_FORMATS format, UUID versionNo) {
+        return new BlobMetaData(onDisk, lobLocation, lastUpdatedTime, format, versionNo);
+    }
+
+    /**
+     * Stores metadata about the BLOB
+     */
+    public static final class BlobMetaData implements Serializable {
+
```

Comment:
Given the blob api prestoring a size here would be nice.

Reply:
Will do.